server/src/server.common.ts

```diff
@@ -181,53 +177,111 @@ export function startServer(connection: _Connection) {
 		connection.sendDiagnostics({ uri: textDocument.uri, diagnostics });
 	}
 
-	function validateYAML(textDocument: TextDocument): void {
-		connection.sendDiagnostics({
-			uri: textDocument.uri,
-			diagnostics: [],
-		});
-
+	function validateYamlSyntaxAndModel(textDocument: TextDocument): Diagnostic[] {
 		const diagnostics: Diagnostic[] = [];
 
 		const lineCounter = new LineCounter();
-		const doc = parseDocument(textDocument.getText(), {
+		const yamlDoc = parseDocument(textDocument.getText(), {
 			lineCounter,
 			keepSourceTokens: true,
-			uniqueKeys: false
 		});
 
+		const map = new YAMLSourceMap();
+		map.doMap(yamlDoc.contents);
+
 		// Basic syntax errors
-		for (const err of doc.errors) {
+		for (const err of yamlDoc.errors) {
 			diagnostics.push({ message: err.message, range: rangeFromLinePos(err.linePos) });
 		}
 
-		// Get location of model in CST
-		if (doc.has("model")) {
-			let position: {line: number, col: number};
-
-			// Get the model token and find its position
-			(doc.contents?.srcToken as BlockMap).items.forEach(i => {
-				if (i.key?.offset !== undefined && (i.key as SourceToken).source === "model") {
-					position = lineCounter.linePos(i.key?.offset);
-				}
-			});
+		// If no diagnostics, continue parsing.
+		if (diagnostics.length === 0) {
+
+			const validator = new Ajv();
+			const validate = validator.compile(openfgaYaml);
+
+			const valid = validate(yamlDoc.toJSON());
+			if (!valid) {
+				validate.errors?.forEach((e: ErrorObject) => {
+					console.error(JSON.stringify(e as ErrorObject, null, 2));
+
+					let start = { line: 0, character: 0 };
+					let end = { line: 0, character: 0 };
+					let message;
+
+					if (e.keyword === "oneOf") {
```

Comment:
Can we do better? Do we have any other signals about the context other than `oneOf`?

I'm asking because `tuple_file` will be a thing very soon

Reply:
```
oneOf: [
  { required: ["model_file"] },
  { required: ["model"] }
],
properties: {
  ...
},
```

The error I get back when both are assigned;
```
{
  instancePath: '',
  schemaPath: '#/oneOf',
  keyword: 'oneOf',
  params: { passingSchemas: [ 0, 1 ] },
  message: 'must match exactly one schema in oneOf'
}
```

Missing both fields gets;
```
{
  "instancePath": "",
  "schemaPath": "#/oneOf",
  "keyword": "oneOf",
  "params": {
    "passingSchemas": null
  },
  "message": "must match exactly one schema in oneOf"
```

Reply:
OK, so if the other oneOf is in tuples, it would have a different path right? Can we confirm that?

Reply:
> Can we confirm that?

Happy to investigate, do you have a sample for how the YAML will look with `tuple_file`?

Reply:
For example:
```yaml
name: Store Name # store name, optional
model_file: ./model.fga # a global model that would apply to all tests
tuples_file: ./some_tuples.yaml
tests: # required
  - name: test-1
    description: testing that the model works # optional
    tuples_file: ./some_other_tuples.json
    check: # a set of checks to run
      - user: user:anne
        object: folder:1
        assertions:
          # a set of expected results for each relation
          can_view: true
          can_write: true
          can_share: false
      - user: user:beth
        object: folder:2
        assertions:
          # a set of expected results for each relation
          can_view: true
          can_write: false
          can_share: false
    list_objects: # a set of list objects to run
      - user: user:anne
        type: folder
        assertions:
          # a set of expected results for each relation
          can_view:
            - folder:1
            - folder:2
          can_write:
            - folder:1
            - folder:2
          can_share: []
      - user: user:beth
        type: folder
        assertions:
          # a set of expected results for each relation
          can_view:
            - folder:2
          can_write: []
          can_share: []
```

Reply:
Cant think of a great way of doing this but here is what is working:

```
OPENFGA_YAML_SCHEMA = {
	type: "object",
	oneOf: [
		{ required: ["model_file", "tuples_file"] },
		{ required: ["model", "tuples_file"] },
		{ required: ["model_file", "tuples"] },
		{ required: ["model", "tuples"] },
	],
```

Super unhappy with this config as it scales terribly.

For the test tuples this works:
```
oneOf: [
  { required: ["tuples_file"] },
  { required: ["tuples"] }
],
```				

The error messaging here is not helpful for marking a particular problem.

Reply:
After some testing, this config produces some very misleading error messages. The logic is enforced correctly, but the messaging is very confusing.